File: src/Client/AuthenticationProviders/Basic.php
<?php

namespace HaveAPI\Client\AuthenticationProviders;

/**
 * Provider for HTTP basic authentication.
 * 
 * It accepts `username` and `password` as options.
 */
class Basic extends Base {
	public function authenticate($request) {
		$request->authenticateWith($this->opts['username'], $this->opts['password']);
	}
}

File: lib/haveapi/client/authentication/basic.ex
defmodule HaveAPI.Client.Authentication.Basic do
  alias HaveAPI.Client

  @behaviour Client.Authentication

  def setup(_conn, opts) do
    %{username: opts[:username], password: opts[:password]}
  end

  def authenticate(req, opts) do
    Client.Request.add_header(
      req,
      "Authorization",
      "basic " <> Base.encode64("#{opts.username}:#{opts.password}")
    )
  end
end
